### ISSUE TITLE
Double check why DeepDiff is unable to build in Xcode 9 in Mini Bot 

### ISSUE THREAD

**Issue by NghiaTranUIT** [toggl/toggldesktop#2752]
### 💻 Environment
Platform: macOS
OS Version:  Mini bot
Toggl Version: Latest master (after merge DeepDiff)

### 🐞 Actual behavior
- Unable to build DeepDiff

### 💯 Expected behavior
- DeepDiff should be able to be archived successfully.

### 🔨 Steps to reproduce

1. Checkout master
2. Run on Xcode 9 from Mini Bot
3. 💥 

Related to #2728

**Comment by NghiaTranUIT** [comment]
Technically, DeepDiff is written on Swift 4.1, which introduce [CompactMap](https://useyourloaf.com/blog/replacing-flatmap-with-compactmap/) func.

Only >= Xcode 9.3 is bundled. Consequently, Xcode 9.0 on Mini 🤖  is unable to archive it.

However, Xcode 9.3 require >= macOS 10.13.

## Solution
### 1. Upgrade Xcode 10 and macOS 10.13.6
#### Pros 
- Mini bot is upgrade latest OS
- Deal with #2728 easier, since Xcode 10 support dark theme "automatically"

#### Cons
- More effort
- Some build, code-sign scripts could be broken.

### 2. Keep Xcode 9.0 and macOS 10.12 as it is
We modify DeepDiff directly, by replacing all `CompactMap`, with `FlatMap`, since it's same technically.

There 5 places we should replace
<img width="324" alt="screen shot 2019-01-10 at 14 15 52" src="https://user-images.githubusercontent.com/5878421/50952214-47b94780-14e2-11e9-9e37-d7994acd489c.png">

#### Pros 
- Less work
- Keep mini bot as it is

#### Cons
- Impossible to tackle Dark theme #2728

**Comment by IndrekV** [comment]
Lets upgrade. I'll check if the building with SDK 10.11 works ok with Xcode 10 if it's ok we can go ahead and start upgrading.

**Comment by NghiaTranUIT** [comment]
Got it. Let me know when you finish it. I will start upgrading 👍

**Comment by IndrekV** [comment]
The swift part will not compile because the macos sdk 10.11 we use does not support it. In that sense it makes no difference if we have Xcode 9 or Xcode 10. We need to somehow test if when building with Xcode 10 and sdk 10.14 the app works on macos 10.11. If I remember correctly some part of the signing logic was changed from version 10.12 that made the app fail at startup on older macos-s

If version built with sdk 10.14 works on macOS 10.11 we can go ahead and upgrade. 
If not we have to come up with another plan.

Probably move to objective-c with the diff solution or see if we can move the supported version of macos to 10.12

**Comment by NghiaTranUIT** [comment]
I agree. I will give it a try next Monday by installing virtual OS 10.11 on Parallel. Let see if it works or now 👍

**Comment by NghiaTranUIT** [comment]
If you guys wonder how to install macOS 10.11 effortlessly. 

Here is the instruction 📖 

1. Install Parallels app.
2. Download macOS 10.11 DMG file.
3. Mount macOS 10.11.dmg
4. Open Parallels, and select "Install Windows or another OS from Image from DVD or Image file"
5. Select "Image file" -> Select a file
6. Select option **All file** then selecting the "Install OS X El Captain.app" in this mounted disk
7. At this time, Paralles is able to recognize the OS X.
8. Continue installing as usual.

### Notes:
1. We don't need to convert DMG -> ISO anymore.
2. If we select "DMG" file directly in 5th step. Parallels is unable to proceed due to "Un-recognize the OS" 💥

**Comment by NghiaTranUIT** [comment]
Here is the result after the experiment, which is 

> Is TogglDesktop, which is built from OS 10.14, Xcode 10 and Minimum Deployment Target = 10.11, able to run on OS 10.11?

## TL;DR
Yes, it's able to run smoothly on OS 10.11

## Detail results
### Build Platform
- macOS: 10.14.2
- Xcode 10.1
- Commit: https://github.com/toggl/toggldesktop/commit/9f9799e2bada96d3cbac09fe1b70712e097ade4f on master
- Remove `$(XCODELEGACYSDK)` in make file
- Don't need to copy `macOS10.11.sdk` to Xcode's content

### Target platform
- Fresh install macOS 10.11 on Parallels 

### Attempts
1. We export TogglDesktop by archiving from Product Menu -> Archive -> Export to development team.
2. or Run `make app`

### Result
- [x] Both attempts works perfectly on Target Platform.
- [x] The dynamic library linker is correct (see Additions)
- [x] Icon is remained.

![screen shot 2019-01-14 at 3 58 41 pm](https://user-images.githubusercontent.com/5878421/51104707-689ed700-1819-11e9-9eaf-b91fdf5618e6.png)

### Additions
To make sure the linker is okay, I execute both commands
```bash
$ otool -L ~/Application/TogglDesktop.app/Contents/MacOS/TogglDesktop
$ otool -L ~/Application/TogglDesktop.app/Contents/Frameworks/TogglDesktopLibrary.dylib
```

<details><summary>TogglDeskop Binrary's linker</summary>
<p>
	@executable_path/../../Contents/Frameworks/TogglDesktopLibrary.dylib (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 158.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 58286.220.15)
	/System/Library/Frameworks/ExceptionHandling.framework/Versions/A/ExceptionHandling (compatibility version 1.0.0, current version 13.0.0)
	@executable_path/../../Contents/Frameworks/CrashReporter.framework/Versions/A/CrashReporter (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 963.200.27)
	@rpath/Sparkle.framework/Versions/A/Sparkle (compatibility version 1.6.0, current version 1.21.2)
	@executable_path/../../Contents/Frameworks/TFDatePicker.framework/Versions/A/TFDatePicker (compatibility version 1.0.0, current version 1.0.0)
	@rpath/DeepDiff.framework/Versions/A/DeepDiff (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 23.0.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1560.12.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 1671.10.106)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 50.1.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1560.12.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1247.4.1)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 934.0.0)
	@rpath/libswiftAppKit.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftCore.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftCoreData.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftCoreFoundation.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftCoreGraphics.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftCoreImage.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftDarwin.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftDispatch.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftFoundation.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftIOKit.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftMetal.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftObjectiveC.dylib (compatibility version 1.0.0, current version 1000.11.42)
	@rpath/libswiftQuartzCore.dylib (compatibility version 1.0.0, current version 1000.11.42)
</p>
</details>

<details><summary>TogglDeskopLibrary 's linker</summary>
<p>
	@loader_path/TogglDesktopLibrary.dylib (compatibility version 1.0.0, current version 1.0.0)
	@loader_path/libPocoNet.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoFoundation.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoUtil.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoData.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoCrypto.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoJSON.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoDataSQLite.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoNetSSL.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	@loader_path/libPocoXML.60.dylib (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 158.0.0)
	/System/Library/Frameworks/WebKit.framework/Versions/A/WebKit (compatibility version 1.0.0, current version 606.2.104)
	@loader_path/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	@loader_path/libssl.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 50.1.0)
	/System/Library/Frameworks/CFNetwork.framework/Versions/A/CFNetwork (compatibility version 1.0.0, current version 975.0.3)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1560.12.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1247.4.1)
</p>
</details>

**Comment by IndrekV** [comment]
Nice job @NghiaTranUIT let's go forward with upgrading mini buildbot and Xcode.

**Comment by NghiaTranUIT** [comment]
@IndrekV The upgrade of mini 🤖 to OS 10.14.2 and Xcode 10.1 are done.

I've double check that mini bot is able to archive app as usual (by VNC and run `make app` directly in its terminal, because private_key is unable to access from SSH session)
